### ISSUE TITLE
DNF and thousands of test fails on Linux/Java 11 builds

### ISSUE THREAD

**Issue by iloveeclipse** [eclipse-platform/eclipse.platform.swt#133]
On I20220507-1800 we first time saw a horror picture on Linux / Java 11 tests, now it happened again.
![image](https://user-images.githubusercontent.com/964108/167886907-919c9e62-a5b6-4fe7-bd9d-9594e73f9445.png)

I assume this could be caused by same or similar[ system docker files update](https://github.com/eclipse-platform/eclipse.platform.swt/issues/108#issuecomment-1122190900) we saw in #108.
What is strange here - it happens not every time, but if it happens, lot of tests are broken, but some are also fully OK.

I assume the common root cause could be **also** webkit update and affected tests do something that causes webkit to crash or hang JVM, but I'm not sure, so documenting here the observations.

Following builds were affected so far:
- https://download.eclipse.org/eclipse/downloads/drops4/I20220507-1800/testResults.php 
- https://download.eclipse.org/eclipse/downloads/drops4/I20220510-1800/testResults.php
- - https://ci.eclipse.org/releng/view/Automated%20tests/job/ep424I-unit-cen64-gtk3-java11/88/console
- - https://ci.eclipse.org/releng/view/Automated%20tests/job/ep424I-unit-cen64-gtk3-java11/88/artifact/workarea/I20220510-1800/eclipse-testing/results/consolelogs/ep424I-unit-cen64-gtk3-java11_linux.gtk.x86_64_11_consolelog.txt

**Comment by mickaelistria** [comment]
I think blaming the Docker image is correct. It could be that since last update, the configuration for Xvnc is incorrect and Xvnc is not starting.

**Comment by iloveeclipse** [comment]
> The question is: why it happens not every build and only on Java 11?

If I see it right, this happens on every docker image rebuild here: https://ci.eclipse.org/releng/view/Miscellaneous/job/Build-Docker-images/

![image](https://user-images.githubusercontent.com/964108/167891983-55487097-ab81-4831-81a5-c72a88ef024d.png)

@sravanlakkimsetti : could it be, Java 11 / Linux job is triggered first & runs into some troubles with new docker image for the very first time, which are "fixed" by some magic during the build and so later jobs with java 17 / 18 are fine?

**Comment by iloveeclipse** [comment]
@sravanlakkimsetti : where are docker files used are hosted? I see the build uses
`/home/genie.releng/jenkins/workspace/Build-Docker-images/eclipse.platform.releng.aggregator/cje-production/dockerfiles/`.
Shouldn't they be put into some repository like https://github.com/eclipse-platform/eclipse.platform.releng.aggregator.git so one can see & correlate test fails to docker changes?

**Comment by iloveeclipse** [comment]
> That's where exactly dockerfiles are https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/tree/master/cje-production/dockerfiles . 

OK

> But the fact that you build the same dockerfile doesn't mean you get the same image as the OS updates are applied.

NOK, I assume this is because we just expect "latest greatest" on every build, right?

So can we have an agreement to announce docker files rebuild via ticket on platform releng repo & document the versions generated by the docker image rebuild, so that interesting parties are 1) informed and 2) can diff the results without asking experts where to find what?

**Comment by akurtakov** [comment]
> > That's where exactly dockerfiles are https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/tree/master/cje-production/dockerfiles .
> 
> OK
> 
> > But the fact that you build the same dockerfile doesn't mean you get the same image as the OS updates are applied.
> 
> NOK, I assume this is because we just expect "latest greatest" on every build, right?

This is because we want to have tests as close to what users would have as possible. Everytime, we see an issue that could origin in the OS we have to be sure that it's run with all system updates installed, otherwise one is told to do so by every OS vendor. 

> 
> So can we have an agreement to announce docker files rebuild via ticket on platform releng repo & document the versions generated by the docker image rebuild, so that interesting parties are 1) informed and 2) can diff the results without asking experts where to find what?

I would leave it to @sravanlakkimsetti to comment on how and if this is feasible. 
FWIW, this process worked fine for years so we may be seeing a glitch.

**Comment by iloveeclipse** [comment]
> It could be that since last update, the configuration for Xvnc is incorrect and Xvnc is not starting

OK, I see in [`eclipse.platform.releng.aggregator/cje-production/dockerfiles/centos-gtk3-metacity/8-gtk3/scripts/xstartup_metacity.sh`](https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/blob/b6181eba9b8dcb74275ac251937c135dce27d36e/cje-production/dockerfiles/centos-gtk3-metacity/8-gtk3/scripts/xstartup_metacity.sh#L6)
following:
`Xvnc ${DISPLAY} -geometry 1440x900 -depth 16 -dpi 100 -PasswordFile ${HOME}/.vnc/passwd &`
and in log
`unable to create directory '/home/vnc/.cache/dconf': Permission denied`.
I see you've tried to address `/home/vnc/` write issue in https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/pull/259.

We will see if that helps here too.

> I think we should revert earlier patch to Dockerfiles as the current state is worse than previous one; and then work on a better solution.

Not sure which one you means: the  https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/pull/259 was not **yet** used for SDK build.

**Comment by iloveeclipse** [comment]
> Not sure which one you means: the https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/pull/259 was not yet used for SDK build.

Scratch this, checking into git AFTER SDK build doesn't mean the image wasn't built **before**. 

This is exceptionally surprising process.

**Comment by iloveeclipse** [comment]
> I've retriggered the test run to see whether the results are one time or point to bigger problem 

I would guess this will be "fine" (with usual test fails but without DNF's), because the pattern I see is that the Java 11 build has DNF issues only directly after docker image rebuild.

**Comment by akurtakov** [comment]
Could this be related to the recent Jenkins LTS update on JIPPs? So docker images "fetching" has issues.

**Comment by mickaelistria** [comment]
> This build was done before your patch, right?

Yes I think so. Recent changes in webkit-gtk broke the AllBrowserTests, I tried to implement some adaptation for those to work; and that caused even more failures because it now looks like Xvnc isn't started at all which I think is caused by a regression from my change.

**Comment by iloveeclipse** [comment]
But https://download.eclipse.org/eclipse/downloads/drops4/I20220507-1800/testResults.php shows same DNF fails, so how you patch could make it worse?

**Comment by iloveeclipse** [comment]
> I've retriggered the test run to see whether the results are one time or point to bigger problem https://ci.eclipse.org/releng/view/Automated%20tests/job/ep424I-unit-cen64-gtk3-java11/89/

Test Result (3,652 failures / ±0)

**Comment by sravanlakkimsetti** [comment]
We don't see this problem today. 

> I think blaming the Docker image is correct. It could be that since last update, the configuration for Xvnc is incorrect and Xvnc is not starting.
I doubt this is related to your change to provide access to $HOME. We have see this problem with https://download.eclipse.org/eclipse/downloads/drops4/I20220507-1800/testResults.php. 

We use the same docker image in all test runs including smoke test runs. Interestingly platform UI test run used by smoke tests ran successfully https://ci.eclipse.org/releng/view/Smoke%20Tests/job/Start-smoke-tests/1227/

**Comment by sravanlakkimsetti** [comment]
the vnc connection seems to have been broken during org.eclipse.jface.tests.databinding.BindingTestSuite run

**Comment by sravanlakkimsetti** [comment]
This could be a problem from foundation side.

**Comment by sravanlakkimsetti** [comment]
> So can we have an agreement to announce docker files rebuild via ticket on platform releng repo & document the versions generated by the docker image rebuild, so that interesting parties are 1) informed and 2) can diff the results without asking experts where to find what?

We do rebuild our docker images to avoid vulnerabilities. We do rebuild 3 times a week. Foundation also does that for all the images they use(centos-7, centos-8, base etc). 

One difference between foundation images and us is our images are based on centos stream vs archived centos images. We end up identifying most of the problems early in our tests.

**Comment by sravanlakkimsetti** [comment]
I made a change in the job to fetch java 11 from openjdk and use it test run. This is similar to java 17 and java 18 runs. Till today we were using installed java from foundation.

**Comment by akurtakov** [comment]
I think we are done here unless someone has a better workflow for keeping Docker images recent and decent and time to implement it. In that case please reopen with the plan.

**Comment by iloveeclipse** [comment]
> Now we are using java 11 from openjdk instead of adoptopenjdk
> I think we are done here

So the conclusion is that the problem is in java packages used? Do we have understanding what is wrong on adoptopenjdk, why that isn't "good enough" for us, or it is just a guess because that's the only difference with Java17/18 builds?

>  unless someone has a better workflow for keeping Docker images recent and decent and time to implement it. In that case please reopen with the plan.

I think the only thing I personally miss right now is the transparency in the docker rebuilds. I think sending a mail to releng list from a job should be doable? I will open a task.

**Comment by akurtakov** [comment]
@iloveeclipse "From now there should be a notification in platform-releng-dev mailing list." so @sravanlakkimsetti already did that.

**Comment by sravanlakkimsetti** [comment]
> So the conclusion is that the problem is in java packages used?

Not necessarily. I have see multiple failures like this in the past all related to stability of VNC connectivity. There are multiple scenarios here
1. Port numbers used may get banned because of security policies(happened multiple times)
2. VNC can disconnect after no data transmission for certain amount of time. Basically timeout. (probably this might be the reason here).

Probably second case is the likely scenario here. If this the case we need to figure exactly which operation is taking more time to respond, causing vnc connection to timeout.

**Comment by iloveeclipse** [comment]
Happened again in https://download.eclipse.org/eclipse/downloads/drops4/I20220512-1800/testResults.php after docker rebuild via https://ci.eclipse.org/releng/job/Build-Docker-images/366//console.

**Comment by akurtakov** [comment]
@mbarbero @fgurr do you have any idea? It seems to happen only on very first run after docker image rebuild and started to happen last week. It somehow correlates with the Jenkins LTS update time in my mind. Could you check whether there is some issue with image initially or smth similar?

**Comment by sravanlakkimsetti** [comment]
We are seeing VNC disconnection. Is there is a timeout for VNC connection? if so can this be increased?

**Comment by fredg02** [comment]
> It seems to happen only on very first run after docker image rebuild and started to happen last week. It somehow correlates with the Jenkins LTS update time in my mind.

Jenkins LTS upgrade was done on May 9, the issues started to happen before (at least May 7, AFAICT). So it seems unrelated.

There was another change related to Docker (switching from podman to docker-ce on https://ci.eclipse.org/releng/computer/zyt5z%2Dcentos83/), but that happened on April 26, so issues should have appeared earlier.

**Comment by fredg02** [comment]
> We are seeing VNC disconnection. Is there is a timeout for VNC connection? if so can this be increased?

Please point me to the failing build.

**Comment by akurtakov** [comment]
@fredg02 here it is https://ci.eclipse.org/releng/job/ep424I-unit-cen64-gtk3-java11/93/

**Comment by fredg02** [comment]
```
20:57:07 sh: line 28: 43 Killed "Xvnc" ":53" "-geometry" "1024x768" "-depth" "24" "-ac" "-SecurityTypes" "none" "-noreset"
```

The `Killed` keyword is a good indication of an OutOfMemory error, where the resource limits caused the process to be killed. So I'd suggest to increase the memory resource limits (e.g. to 6GB) to see if that makes a difference.

In order to avoid misleading error messages, I'd recommend to add missing applications to the docker image (e.g. `pkill`, `hostname` and `ps`).

**Comment by iloveeclipse** [comment]
> So I'd suggest to increase the memory resource limits (e.g. to 6GB) to see if that makes a difference.

Thanks for the hint. **How** and where one can do that?

**Comment by fredg02** [comment]
> Thanks for the hint. How and where one can do that?

In the Jenkins pipeline
```
  - name: "custom"
    image: "eclipse/platformreleng-centos-gtk3-metacity:8"
    imagePullPolicy: "Always"
    resources:
      limits:
        memory: "4096Mi"   <= change to "6144Mi"
        cpu: "2000m"
      requests:
        memory: "512Mi"
        cpu: "1000m"
```
here: https://ci.eclipse.org/releng/job/ep424I-unit-cen64-gtk3-java11/configure

**Comment by iloveeclipse** [comment]
Again everything failed in [I20220513-1800](https://download.eclipse.org/eclipse/downloads/drops4/I20220513-1800) / https://ci.eclipse.org/releng/view/Automated%20tests/job/ep424I-unit-cen64-gtk3-java11/95/

I've changed the job config now to use more memory and restarted the build.

**Comment by iloveeclipse** [comment]
@fredg02 : let's hope that was it!
https://ci.eclipse.org/releng/view/Automated%20tests/job/ep424I-unit-cen64-gtk3-java11/96/ is green after increasing memory.

Questions now: **if** the memory was the issue:
1) Why did we observe that only for Java 11 / Linux build, but not for Linux builds using save docker image but with java 17 / Java 18?
2) What changed in system on I20220507-1800 that caused this? Are some tests requiring more memory now?
3) This issue here timely correlates with fail in AllBrowsertests in https://github.com/eclipse-platform/eclipse.platform.swt/issues/108, which is however **still broken** on all Linux builds. Is this just co-incident or is the problem caused by updated webkit ? ( which might be requires more memory ??? ).

**Comment by iloveeclipse** [comment]
Still no crashes after memory increase, so let's close.

Would be great if anyone could answer to my previous question.

**Comment by akurtakov** [comment]
@iloveeclipse so you don't think your question is ignored - at least from our side no one investigated the issue so we can't give an answer.

**Comment by fredg02** [comment]
> @fredg02 You've show some informations about some particular pod memory consumptions in the past; would it be possible to get information from a 11 days old pod whose Jenkins build was garbage collected?

Nope, as you can imagine, storing the data of all our builds for 11 days or longer would required significant resources.

> @fredg02 is there an easy way to get this very useful RAM consumption graph for a given pod without pinging you every 2 seconds? Maybe drilling down on a per-process consumption and see what seems to consume more RAM than expected...

Not yet. While we eventually want to make more metrics available to the public, day-to-day tasks keep pushing this effort down the stack. Retention time will definitely be limited (see above).

If memory leaks are suspected, I'd recommend to try to reproduce this locally with proper tooling. This will always be faster and more accurate than what we can provide.
